src/Microsoft.AspNetCore.ResponseCompression/ResponseCompressionProvider.cs

```diff
@@ -62,42 +72,79 @@ public ResponseCompressionProvider(IServiceProvider services, IOptions<ResponseC
         /// <inheritdoc />
         public virtual ICompressionProvider GetCompressionProvider(HttpContext context)
         {
-            IList<StringWithQualityHeaderValue> unsorted;
-
             // e.g. Accept-Encoding: gzip, deflate, sdch
             var accept = context.Request.Headers[HeaderNames.AcceptEncoding];
-            if (!StringValues.IsNullOrEmpty(accept)
-                && StringWithQualityHeaderValue.TryParseList(accept, out unsorted)
-                && unsorted != null && unsorted.Count > 0)
+
+            if (StringValues.IsNullOrEmpty(accept))
             {
-                // TODO PERF: clients don't usually include quality values so this sort will not have any effect. Fast-path?
-                var sorted = unsorted
-                    .Where(s => s.Quality.GetValueOrDefault(1) > 0)
-                    .OrderByDescending(s => s.Quality.GetValueOrDefault(1));
+                return null;
+            }
 
-                foreach (var encoding in sorted)
+            if (StringWithQualityHeaderValue.TryParseList(accept, out var encodings))
+            {
+                if (encodings.Count == 0)
                 {
-                    // There will rarely be more than three providers, and there's only one by default
-                    foreach (var provider in _providers)
+                    return null;
+                }
+
+                var candidates = new HashSet<ProviderCandidate>();
+
+                foreach (var encoding in encodings)
+                {
+                    var encodingName = encoding.Value;
+                    var quality = encoding.Quality.GetValueOrDefault(1);
+
+                    if (Math.Abs(quality) < double.Epsilon)
+                    {
+                        continue;
+                    }
+
+                    for (int i = 0; i < _providers.Length; i++)
                     {
-                        if (StringSegment.Equals(provider.EncodingName, encoding.Value, StringComparison.OrdinalIgnoreCase))
+                        var provider = _providers[i];
+
+                        if (StringSegment.Equals(provider.EncodingName, encodingName, StringComparison.OrdinalIgnoreCase))
                         {
-                            return provider;
+                            candidates.Add(new ProviderCandidate(provider.EncodingName, quality, i, provider));
```

Comment:
I don't think it's much more work, it's just doing it the other way around by matching first and ordering after, but we can certainly try to be smarter to avoid some work 👍🏻

Reply:
Might be worth checking/considering making _providers a Dictionary? Besides that, I don't see much you can do.